src/app/clusters/zone-management-server/zone-geometry.h

```diff
@@ -0,0 +1,335 @@
+        {
+            TwoDCartesianVertexStruct p1 = vertices[i];
+            TwoDCartesianVertexStruct q1 = vertices[(i + 1) % vertexCount];
+
+            // j starts from i+2, skipping the adjacent vertex i+1.
+            for (size_t j = i + 2; j < vertexCount; ++j)
+            {
+                TwoDCartesianVertexStruct p2 = vertices[j];
+                TwoDCartesianVertexStruct q2 = vertices[(j + 1) % vertexCount];
+
+                // Skip segments that share an endpoint (i.e., adjacent segments)
+                if ((j + 1) % vertexCount == i)
+                {
+                    continue;
+                }
+
+                if (DoSegmentsIntersect(p1, q1, p2, q2))
+                {
+                    ChipLogDetail(Zcl, "Zone segment[(%u,%u),(%u,%u)] intersects with segment[(%u,%u),(%u,%u)]", p1.x, p1.y, q1.x,
+                                  q1.y, p2.x, p2.y, q2.x, q2.y);
+                    return true; // Found a self-intersection
+                }
+            }
+        }
+
+        return false; // No self-intersection found
+    }
+
+private:
+    // Helper function: Check if point q lies on segment pr (assuming p, q, r are collinear)
+    // Checks if q is within the bounding box defined by p and r.
+    // Returns true if the coordinates of q are greater than the min and smaller
+    // than the max of the corresponding coordinates of p and r.
+    static bool OnSegment(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                          const TwoDCartesianVertexStruct & r)
+    {
+        return q.x <= std::max(p.x, r.x) && q.x >= std::min(p.x, r.x) && q.y <= std::max(p.y, r.y) && q.y >= std::min(p.y, r.y);
+    }
+
+    // Helper function: Determine the orientation of the ordered triplet (p, q, r) of vertices.
+    // The purpose of this function is to determine if 'r' is to the left, right
+    // of the vector 'pq', or is collinear with it.
+    // Employs the cross-product computation (determinant of the coordinate matrix) of
+    // vectors pq and pr to find the direction of the Z axis. The rotation
+    // considered is from vector pq to pr.
+    // The sign of the determinant is used to infer the direction of the cross-product vector
+    // and hence the orientation of the 3 coordinates. A value of 0 indicates that the points
+    // are collinear.
+    // Using the right-hand rule Z > 0 means pointing upward from the 2D
+    // plane (counter-clockwise rotation) and Z < 0 means pointing downward into
+    // the 2D plane (clock-wise rotation).
+    // 0 --> p, q and r are collinear
+    // 1 --> Counterclockwise
+    // 2 --> Clockwise
+    static OrientationEnum Orientation(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                                       const TwoDCartesianVertexStruct & r)
+    {
+        long long val = static_cast<long long>((q.x - p.x) * (r.y - p.y)) - static_cast<long long>((q.y - p.y) * (r.x - p.x));
```

Comment:
What are the static_casts doing here?  They don't make sense.... If the worry is possible overflow on the multiplication, then you need to cast _before_ multiplying, no?